adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -287,47 +261,175 @@ def _final_callback(self):
+    scale calculations. Its subclasses can be used in combination with
+    ``adaptdl.torch.parallel.AdaptiveDataParallel`` and ``torch.optim.SGD``.
+
+    .. code-block:: python
+
+        optim = torch.optim.SGD(model, lr=0.001)
+        adascale = AdaScale()
+        model = AdaptiveDataParallel(model, optim, adascale)
+
+        for epoch in ...:
+            for batch in ...:
+                optim.zero_grad()
+                loss = ...
+                loss.backward()
+                adascale.step()
+    """
+    def __init__(self):
+        # instance of AdaptiveDataParallel, needs to be set before any of the
+        # methods can be used
+        self.adp = None
+
+    def calculate_lr_factors(self, scale):
+        raise NotImplementedError
+
+    def step(self, *args, **kwargs):
+        """
+        Run one optimizer step. Essentially just invokes
+        ``optimizer.step(*args, **kwargs)`` with a scaled learning rate.
+
+        Arguments:
+            args: Positional arguments passed to ``optimizer.step``.
+            kwargs: Keyword arguments passed to ``optimizer.step``.
+        """
+        if not self.adp:
+            raise ValueError("AdaptiveDataParallel instance is not set!")
+        if not self.adp.require_backward_grad_sync:
+            return
+        self.adp.gns.run_step(self.calculate_lr_factors, *args, **kwargs)
 
     def patch_optimizer(self):
         """
-        Monkey-patch the optimizer's step function with :meth:`AdaScale.step`.
+        Monkey-patch the optimizer's step function with
+        :meth:`ScalingRuleBase.step`.
         """
         # TODO: detect if the optimizer has already been patched.
+        self.adp.gns.patch_optimizer(self.step)
 
-        @functools.wraps(self._optimizer.step)
-        def step_wrapper(optim, *args, **kwargs):
-            return self.step(*args, **kwargs)
 
-        @functools.wraps(self._optimizer.zero_grad)
-        def zero_wrapper(optim, *args, **kwargs):
-            return self.zero_grad(*args, **kwargs)
-        self._optimizer.step = MethodType(step_wrapper, self._optimizer)
-        self._optimizer.zero_grad = MethodType(zero_wrapper, self._optimizer)
+class AdaScale(ScalingRuleBase):
+    """
+    Implements the AdaScale_ algorithm for scaling the learning rate for
+    distributed and large batch size training.
+
+    .. _AdaScale: https://proceedings.icml.cc/static/paper_files/icml/2020/4682-Supplemental.pdf
+    """  # noqa: E501
+
+    def calculate_lr_factors(self, scale):
+        """Calculate factors to be applied to lr for each parameter group."""
+        var = self.adp.gns.state["var_avg"]
+        sqr = self.adp.gns.state["sqr_avg"]
+        var = np.maximum(var, 1e-6)
+        sqr = np.maximum(sqr,  0.0)
+        return (var + sqr) / (var / scale + sqr)
+
+
+class LinearScale(ScalingRuleBase):
+
+    def calculate_lr_factors(self, scale):
+        return np.full(len(self.adp.gns.optimizer.param_groups), scale)
```

Comment:
Better to make `optimizer` a `@property` of `AdaptiveDataParallel` than to access it as a field of `gns`